store/sqlstore/migration_create_index.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sqlstore
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+// CreateIndex - asynchronous migration that adds an index to table
+type CreateIndex struct {
+	name      string
+	table     string
+	columns   []string
+	indexType string
+	unique    bool
+}
+
+func NewCreateIndex(indexName string, tableName string, columnNames []string, indexType string, unique bool) *CreateIndex {
+	return &CreateIndex{
+		name:      indexName,
+		table:     tableName,
+		columns:   columnNames,
+		indexType: indexType,
+		unique:    unique,
+	}
+}
+
+// Name returns name of the migration, should be unique
+func (m *CreateIndex) Name() string {
+	return "add_index_" + m.name
+}
+
+// GetStatus returns if the migration should be executed or not
+func (m *CreateIndex) GetStatus(ss SqlStore) (AsyncMigrationStatus, error) {
+	if ss.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		_, errExists := ss.GetMaster().SelectStr("SELECT $1::regclass", m.name)
+		// It should fail if the index does not exist
+		if errExists == nil {
+			return AsyncMigrationStatusSkip, nil
+		}
+		if m.indexType == INDEX_TYPE_FULL_TEXT && len(m.columns) != 1 {
+			return AsyncMigrationStatusFailed, errors.New("Unable to create multi column full text index")
+		}
+	} else if ss.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		if m.indexType == INDEX_TYPE_FULL_TEXT {
+			return AsyncMigrationStatusFailed, errors.New("Unable to create full text index concurrently")
+		}
+		count, err := ss.GetMaster().SelectInt("SELECT COUNT(0) AS index_exists FROM information_schema.statistics WHERE TABLE_SCHEMA = DATABASE() and table_name = ? AND index_name = ?", m.table, m.name)
+		if err != nil {
+			return AsyncMigrationStatusUnknown, err
+		}
+		if count > 0 {
+			return AsyncMigrationStatusSkip, nil
+		}
+	}
+	return AsyncMigrationStatusRun, nil
+}
+
+// Execute runs the migration
+// Explicit connection is passed so that all queries run in a single session
+func (m *CreateIndex) Execute(ctx context.Context, ss SqlStore, conn *sql.Conn) (AsyncMigrationStatus, error) {
+	uniqueStr := ""
+	if m.unique {
+		uniqueStr = "UNIQUE "
+	}
+
+	if ss.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		query := ""
+		if m.indexType == INDEX_TYPE_FULL_TEXT {
+			columnName := m.columns[0]
+			postgresColumnNames := convertMySQLFullTextColumnsToPostgres(columnName)
+			query = "CREATE INDEX CONCURRENTLY " + m.name + " ON " + m.table + " USING gin(to_tsvector('english', " + postgresColumnNames + "))"
```

Comment:
Is there a reason we are not adding "IF NOT EXISTS" ? AFAIU, if we don't add this, it will unnecessarily create this index every time the server restarts, isn't it ?

Reply:
Indices must have unique name globally in postgres, per table in mysql. It would fail with "already exists" error, and it is also checked if the index exists before adding migration to queue.

Reply:
The difference is with "if not exists" it issues notice instead of error, and it's supported from postgres 9.5

Reply:
Ooh just a difference of one minor release :) Okay then checking whether it exists or not before adding to queue is good enough I guess.